hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NameNode.java

```diff
@@ -2051,6 +2056,26 @@ synchronized HAServiceState getServiceState() {
     return state.getServiceState();
   }
 
+  /**
+   * Emit Namenode HA service state as an integer so that one can monitor NN HA
+   * state based on this metric.
+   *
+   * @return  0 when not fully started
+   *          1 for active or standalone (non-HA) NN
+   *          2 for standby
+   *          3 for observer
+   *
```

Comment:
I saw HAState has a stopping enum. We won't encounter that state?

Reply:
Searching codebase, it seems we would set a state to STOPPING state only in YARN ResourceManager HA. We are not using that state in HDFS.